std/algebra/twistededwards/bandersnatch/point.go

```diff
@@ -98,41 +107,87 @@ func (p *Point) Double(api frontend.API, p1 *Point, curve EdCurve) *Point {
 	return p
 }
 
+// phi endomorphism √-2 ∈ 𝒪₋₈
+// (x,y) → λ × (x,y) s.t. λ² = -2 mod Order
+func (p *Point) phi(api frontend.API, p1 *Point, curve EdCurve) *Point {
+
+	xy := api.Mul(p1.X, p1.Y)
+	yy := api.Mul(p1.Y, p1.Y)
+	f := api.Sub(1, yy)
+	f = api.Mul(f, curve.endo1)
+	g := api.Add(yy, curve.endo0)
+	g = api.Mul(g, curve.endo0)
+	h := api.Sub(yy, curve.endo0)
+
+	p.X = api.DivUnchecked(f, xy)
+	p.Y = api.DivUnchecked(g, h)
+
+	return p
+}
+
+var scalarDecompositionHint = hint.NewStaticHint(func(curve ecc.ID, inputs []*big.Int, res []*big.Int) error {
+	// TODO: handle properly negative scalars
```

Comment:
scalars can't be "negative" here. Inputs are big.Int , yes , but they are injected by the solver. So they were fr.Element converted to big.Int and will be 0 <= s < q .

Reply:
yes, but at this point (as big.Int) they can be negative, and it seems `sp[0]` is always negative because `lambda` is 253-bit (max bitsize of scalar to decompose). What I did is negate the `sp[0]` before it is injected to guarantee it is the smallest within the range `[0,q[` and then negate the point inside the scalar mul function.

Reply:
I think that's misleading; you can't inject a negative value here, because it's done at proving time, by the solver. So even if in the circuit definition, you "negate" it, it's going to be encoded in the constraint system, and when the solver will solve the constraints, it is going to be a fr.Element before it is a big int.

Reply:
That's right. Also, if we reduce the scalar mod the Bandersnatch prime order (`s-o`), then the resulting scalar will bigger than 128-bit and we can't benefit from `api.ToBinary(s, 128)` and `for` loop of size 128, which is basically what reduces the number of constraints.

Reply:
so as spotted and explained by @ThomasPiellard `sp[0]` is always negative because we always round above the determinant/2. It turns out that this is very suitable inside a circuit as we force `sp[0]` to be always negative. Thus we avoid incorrect reduction modulo `r` in the hint and we make sure of the scalars max bounds. However, this increases the bound from 128-bit to 129-bit costing +19C for a scalar mul.